decoder.go

```diff
@@ -45,19 +45,27 @@ func NewDecoder(r io.Reader, readTimeout time.Duration) *Decoder {
 func (dec *Decoder) Decode() (Event, error) {
 	pub := new(publication)
 	inDecoding := false
+	var timeoutTimer *time.Timer
+	var timeoutCh <-chan time.Time
+	if dec.readTimeout > 0 {
+		timeoutTimer = time.NewTimer(dec.readTimeout)
+		timeoutCh = timeoutTimer.C
+	}
+	resetTimeout := func(reset bool) {
+		if timeoutTimer != nil {
+			if !timeoutTimer.Stop() {
+				<-timeoutCh
+			}
+			if reset {
+				timeoutTimer.Reset(dec.readTimeout)
+			}
+		}
+	}
 ReadLoop:
 	for {
-		var timeoutTimer *time.Timer
-		var timeoutCh <-chan time.Time
-		if dec.readTimeout > 0 {
-			timeoutTimer = time.NewTimer(dec.readTimeout)
-			timeoutCh = timeoutTimer.C
-		}
 		select {
 		case line := <-dec.linesCh:
-			if timeoutTimer != nil {
-				timeoutTimer.Stop()
-			}
+			resetTimeout(true)
```

Comment:
Would it be slightly simpler to do the reset at the bottom of the for loop and use a `defer timeoutTimer.Stop()` to clean up when we exit early?

Reply:
I'm not sure what you mean. The reset is not the same in each case: on this line, we want the timer to be reinitialized so it can fire again - which requires a Stop and _then_ a Reset - but on line 97, we want it to be permanently shut down.